numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1540,62 +1524,34 @@ pack_inner(const char *inptr,
     npy_intp index = 0;
     int remain = n_in % 8;              /* uneven bits */
 
-#if defined NPY_HAVE_SSE2_INTRINSICS && defined HAVE__M_FROM_INT64
+#if NPY_SIMD
     if (in_stride == 1 && element_size == 1 && n_out > 2) {
-        __m128i zero = _mm_setzero_si128();
+        npyv_u8 v_zero = npyv_zero_u8();
         /* don't handle non-full 8-byte remainder */
         npy_intp vn_out = n_out - (remain ? 1 : 0);
-        vn_out -= (vn_out & 1);
-        for (index = 0; index < vn_out; index += 2) {
-            unsigned int r;
-            npy_uint64 a = *(npy_uint64*)inptr;
-            npy_uint64 b = *(npy_uint64*)(inptr + 8);
-            if (order == 'b') {
-                a = npy_bswap8(a);
-                b = npy_bswap8(b);
+        const int vstep = npyv_nlanes_u64;
+        vn_out -= (vn_out & (vstep - 1));
+        for (index = 0; index < vn_out; index += vstep) {
+            // Maximum parallel ability: handle eight 64-bit integers at one time
+            npy_uint64 a[8];
+            for (int i = 0; i < vstep; i++) {
+                a[i] = *(npy_uint64*)(inptr + 8 * i);
             }
-            
-            /* note x86 can load unaligned */
-            __m128i v = _mm_set_epi64(_m_from_int64(b), _m_from_int64(a));
-            /* false -> 0x00 and true -> 0xFF (there is no cmpneq) */
-            v = _mm_cmpeq_epi8(v, zero);
-            v = _mm_cmpeq_epi8(v, zero);
-            /* extract msb of 16 bytes and pack it into 16 bit */
-            r = _mm_movemask_epi8(v);
-            /* store result */
-            memcpy(outptr, &r, 1);
-            outptr += out_stride;
-            memcpy(outptr, (char*)&r + 1, 1);
-            outptr += out_stride;
-            inptr += 16;
-        }
-    }
-#elif defined NPY_HAVE_NEON
-    if (in_stride == 1 && element_size == 1 && n_out > 2) {
-        /* don't handle non-full 8-byte remainder */
-        npy_intp vn_out = n_out - (remain ? 1 : 0);
-        vn_out -= (vn_out & 1);
-        for (index = 0; index < vn_out; index += 2) {
-            unsigned int r;
-            npy_uint64 a = *((uint64_unaligned*)inptr);
-            npy_uint64 b = *((uint64_unaligned*)(inptr + 8));
             if (order == 'b') {
-                a = npy_bswap8(a);
-                b = npy_bswap8(b);
+                for (int i = 0; i < vstep; i++) {
+                    a[i] = npy_bswap8(a[i]);
+                }
```

Comment:
```suggestion
                a = npyv_rev64_u8(a);
```
`npyv_rev64_u8()` defined by #17789

the final loop should be similar to the following
```C
for (; index < vn_out; index += vstep, inptr += npyv_nlanes_u8) {
    npyv_u8 a = npyv_load_u8((const npy_uint8*)inptr);
    if (order == PACK_ORDER_BIG) {
        a = npyv_rev64_u8(a);
    }
    npy_uint64 bb = npyv_tobits_b8(npyv_cmpneq_u8(a, v_zero));
    for (int i = 0; i < vstep; ++i) {
        memcpy(outptr, (char*)&bb + i, 1);
        outptr += out_stride;
    }
}
```
where `order` defiend as enum:
```C
typedef enum {
    PACK_ORDER_LITTLE = 0,
    PACK_ORDER_BIG
} PACK_ORDER;
```
since its better to guarantee to have a constant hint within the callee `pack_bits(PyObject *input, int axis, char order)`
something like that
```C
const PACK_ORDER ordere = order == 'b' ? PACK_ORDER_BIG : PACK_ORDER_LITTLE;
        pack_inner(PyArray_ITER_DATA(it), PyArray_ITEMSIZE(new),
                   PyArray_DIM(new, axis), PyArray_STRIDE(new, axis),
                   PyArray_ITER_DATA(ot), PyArray_DIM(out, axis),
                   PyArray_STRIDE(out, axis), ordere);
```

Reply:
we should also unroll by x4 to provide faster access to the writing stream
<details>
<summary>something like that</summary>

```C
if (out_stride == 1) {
    const int wstep = vstep*4;
    for (; index <= vn_out - wstep; index += wstep, inptr += npyv_nlanes_u8 * 4, outptr += wstep) {
        npyv_u8 v0 = npyv_load_u8((const npy_uint8*)inptr + npyv_nlanes_u8 * 0);
        npyv_u8 v1 = npyv_load_u8((const npy_uint8*)inptr + npyv_nlanes_u8 * 1);
        npyv_u8 v2 = npyv_load_u8((const npy_uint8*)inptr + npyv_nlanes_u8 * 2);
        npyv_u8 v3 = npyv_load_u8((const npy_uint8*)inptr + npyv_nlanes_u8 * 3);

        if (order == PACK_ORDER_BIG) {
            v0 = npyv_rev64_u8(v0);
            v1 = npyv_rev64_u8(v1);
            v2 = npyv_rev64_u8(v2);
            v3 = npyv_rev64_u8(v3);
        }

        npy_uint64 bb0 = npyv_tobits_b8(npyv_cmpneq_u8(v0, v_zero));
        npy_uint64 bb1 = npyv_tobits_b8(npyv_cmpneq_u8(v1, v_zero));
        npy_uint64 bb2 = npyv_tobits_b8(npyv_cmpneq_u8(v2, v_zero));
        npy_uint64 bb3 = npyv_tobits_b8(npyv_cmpneq_u8(v3, v_zero));

        npy_uint64 *ptr64 = (npy_uint64*)outptr;
        if (NPY_SIMD_WIDTH == 16) {
            npy_uint64 bb = bb0 | (bb1 << 16) | (bb2 << 32) | (bb3 << 48);
            ptr64[0] = bb;
        }
        else if (NPY_SIMD_WIDTH == 32) {
            ptr64[0] = bb0 | (bb1 << 32);
            ptr64[1] = bb2 | (bb3 << 32);
        }
        else {
            ptr64[0] = bb0; ptr64[1] = bb1;
            ptr64[2] = bb2; ptr64[3] = bb3;
        }
    }
}
for (; index < vn_out; index += vstep, inptr += npyv_nlanes_u8) {
    npyv_u8 a = npyv_load_u8((const npy_uint8*)inptr);
    if (order == PACK_ORDER_BIG) {
        a = npyv_rev64_u8(a);
    }
    npy_uint64 bb = npyv_tobits_b8(npyv_cmpneq_u8(a, v_zero));
    for (int i = 0; i < vstep; ++i) {
        memcpy(outptr, (char*)&bb + i, 1);
        outptr += out_stride;
    }
}
```

Reply:
also, it would be better if you could add optimization level 3 to `pack_inner()` via `NPY_GCC_OPT_3`.